### ISSUE TITLE
trilinos/package.py ignores command line cxxflags option

### ISSUE THREAD

**Issue by KineticTheory** [LLNL/spack#1953]
In trying to build trilinos with some special compiler options, I noticed that the recipe ignores `cxxflags` set on the spack command line.  In my case, I was trying to build with gcc:

```
$ spack install trilinos~boost~hdf5~hypre~mumps~suite-sparse ^openmpi@1.10.3 \
       cxxflags=\"-D_GLIBCXX_DEBUG -D_GLIBCXX_DEBUG_PEDANTIC\"
```

These extra flags are not applied when building trilinos.  It appears that the recipe ignores this setting.  Can someone confirm this?

**Comment by tgamblin** [comment]
It shouldn't, as those flags would be added by the Spack compiler wrappers _below_ the Trilinos build.    @davydden: have you played with this?

**Comment by davydden** [comment]
no, i have not. I never needed.

**Comment by tgamblin** [comment]
does the trilinos build use the compiler wrappers `spack_cxx` etc?

**Comment by tgamblin** [comment]
it may be that this stuff is getting lost in the CMake build someplace.

**Comment by KineticTheory** [comment]
@tgamblin Trilinos uses the mpi compile wrappers instead of the spack compilers.  So if the `cxxflags` options are only applied to the spack compilers, they may be omitted when mpi compile wrappers are used.

I hacked the recipe to run `make('VERBOSE=1')` and I see compile lines like this:

```
cd /tmp/kellyt/spack-stage/spack-stage-_iCXln/Trilinos-trilinos-release-12-8-1/spack-build/packages/shards/src && /scratch/vendors/spack/opt/spack/linux-redhat7-x86_64/gcc-4.8.5/openmpi-1.10.3-rrla7crr2sdqot4q5m4jb2ry46s4o2f4/bin/mpicxx   -Dshards_EXPORTS -I/tmp/kellyt/spack-stage/spack-stage-_iCXln/Trilinos-trilinos-release-12-8-1/spack-build -I/tmp/kellyt/spack-stage/spack-stage-_iCXln/Trilinos-trilinos-release-12-8-1/spack-build/packages/shards/src -I/tmp/kellyt/spack-stage/spack-stage-_iCXln/Trilinos-trilinos-release-12-8-1/packages/shards/src -std=c++11 -O3 -fPIC   -o CMakeFiles/shards.dir/Shards_CellTopology.cpp.o -c /tmp/kellyt/spack-stage/spack-stage-_iCXln/Trilinos-trilinos-release-12-8-1/packages/shards/src/Shards_CellTopology.cpp
```

**Comment by KineticTheory** [comment]
I hacked the trilinos recipe to add the flags that I needed, but it seems like this should work 'out-of-the-box' for spack.

**Comment by tgamblin** [comment]
@KineticTheory: Sorry I lost track of the MPI wrapper discussion yesterday.  Did you modify the build to use `"CC"` and `"cc"` for the wrappers?  If so did you use absolute paths to the Cray wrappers, or did you use the Spack wrappers, which ultimately call the cray wrappers?

**Comment by davydden** [comment]
mpi wrappers are usually set up to use Spack's compier, so on non-Cray system i would say that flags should be used by default.

**Comment by davydden** [comment]
in `openmpi` this is done through
```
        spack_env.set('OMPI_CC', spack_cc)
        spack_env.set('OMPI_CXX', spack_cxx)
        spack_env.set('OMPI_FC', spack_fc)
        spack_env.set('OMPI_F77', spack_f77)
```

**Comment by tgamblin** [comment]
On the cray that should probably be omitted, and `CC`, `CXX`, etc. should just be set to `spack_cc`, `spack_cxx`, etc.  Loading the MPI module on the Cray makes `cc` behave like `mpicc`.  I wonder if this is related to a bug @mamelara was seeing yesterday.

**Comment by adamjstewart** [comment]
I was under the impression that `cxxflags` is broken. See #1512 and #1532, which is supposed to fix it.

**Comment by KineticTheory** [comment]
@tgamblin My issue wasn't on a Cray, but on a Linux workstation using spack-built openmpi and system gcc (frrom /usr/bin).  I'm working #1827 separately.

@adamjstewart I'm following #1532 and will try it out soon.  Your PR will most likely resolve this issue.

**Comment by pramodk** [comment]
@adamjstewart : what is expected behaviour of command line compiler flags ? 
```bash

 → spack find -v -f -l mypkg %pgi           #install normally, no flags
==> 1 installed packages.
-- linux-redhat6-x86_64 / pgi@16.5-0 ----------------------------
2bw2hr6 mypkg@develop%pgi

→  spack install --dirty mypkg %pgi ^mpich cflags=\"-O3\" cxxflags=\"-O3\"
==> mypkg is already installed in /somepath/spack/opt/spack/linux-redhat6-x86_64/pgi-16.5-0/mypkg-develop-2bw2hr6bmlqgjr7nphy2qd537mns2qor   #Why?

→  spack install --dirty mypkg %pgi ^mpich cflags=\"-O1\" cxxflags=\"-O2\"
==> mypkg is already installed in /somepath/spack/opt/spack/linux-redhat6-x86_64/pgi-16.5-0/mypkg-develop-2bw2hr6bmlqgjr7nphy2qd537mns2qor  #Why?

```
I haven't specified any flags in `compilers.yaml`. Confusing day for me!

**Comment by tgamblin** [comment]
@becker33

**Comment by tgamblin** [comment]
Per @pramodk this is fixed by #2227.  @KineticTheory: Can you verify

**Comment by pramodk** [comment]
also tested adding `flags` in `compilers.yaml` : 

```
# flags in compilers.yaml
    flags:
      cflags: -O3 -fPIC
      cxxflags: -O3 -fPIC
      cppflags: -O3 -fPIC

# spack install  .... cflags=\"-O1 -g\"

# installed package, cflags are added which is expected behaviour

gypb75b mypkg@develop%gcc cflags="-O3 -O1 -g -fPIC" cppflags="-O3 -fPIC" cxxflags="-O3 -fPIC"
```

**doesn't** trigger #2123 as reported earlier. So all looks good!

**Comment by alalazo** [comment]
> gypb75b mypkg@develop%gcc cflags="-O3 -O1 -g -fPIC" ...

@tgamblin @pramodk `cflags` ordered this way seems a bit weird at a first glance. Why are the `-O1 -g` flags between `-O3` and `-fPIC`? I am worried that if we really pass things unordered to the compiler, we cannot know which option  among a set of mutually exclusive ones have precedence.

**Comment by pramodk** [comment]
> cflags ordered this way seems a bit weird at a first glance

completely agree, and that came to my mind when I first looked at it! And then I thought might be just print time by `spack find`. Let me verify what happens during compilation...I assume these flags shown in `speck-build.out or speck-build.env`?

**Comment by alalazo** [comment]
@pramodk 

>  I assume these flags shown in speck-build.out or speck-build.env?

In `.spack/build.out` you should have a log of all the compiler invocations done in the build. I think we can infer what was the real order from there.

**Comment by pramodk** [comment]
Ok, to confirm : 

```bash
#compilers.yaml has : 

cflags: -O3 -fPIC

#installed package as:

spack install -v mypkg %gcc cflags=\"-O1 -g\"

# and .spack/build.env has

SPACK_CFLAGS=-O3 -O1 -g -fPIC
```

I don't see flags in `.spack/build.out` and I assume that's because those flags are added into spack compiler wrappers.

**Comment by alalazo** [comment]
>I don't see flags in .spack/build.out and I assume that's because those flags are added into spack compiler wrappers

@pramodk  you're right :smile:

**Comment by adamjstewart** [comment]
I believe this issue was fixed by #1532 and #2227. Reopen if I'm wrong.
